client/messaging/message-records.ts

```diff
@@ -1,75 +1,37 @@
 import { Record } from 'immutable'
+import { ChatMessage } from '../chat/chat-message-records'
+import { LobbyMessage } from '../lobbies/lobby-message-records'
 
-export enum ChatMessageType {
+/**
+ * NOTE(2Pac): A common message type that's used in all messaging-related services (e.g. chat,
+ * whispers, lobbies, parties, etc.). For now that's only a default text message, but there might be
+ * more in the future. All other message types that are specific to a particular service are defined
+ * in their respective folders.
+ */
+export enum CommonMessageType {
   TextMessage = 'message',
-  JoinChannel = 'joinChannel',
-  LeaveChannel = 'leaveChannel',
-  NewChannelOwner = 'newOwner',
-  SelfJoinChannel = 'selfJoinChannel',
-  UserOnline = 'userOnline',
-  UserOffline = 'userOffline',
 }
 
 /**
- * The base fields for all chat messages. Any added messages should implement this.
+ * The base fields for all messages. Any added messages should implement this.
  */
-interface BaseChatMessage {
+export interface BaseMessage {
   readonly id: string
   readonly type: string
   readonly time: number
 }
 
 // TODO(tec27): Write a function or something to declare just the extra parts + do the correct
 // typing of the type field automatically.
-export class TextMessage
+export class TextMessageRecord
   extends Record({
     id: '',
-    type: ChatMessageType.TextMessage as typeof ChatMessageType.TextMessage,
+    type: CommonMessageType.TextMessage,
```

Comment:
Does doing it this way break type discrimination? i.e. if you check that `msg.type === CommonMessageType.TextMessage`, does TS then know that `msg` is a `TextMessageRecord`?

If that still works (and works for the Chat/Lobby/Whisper message types too), cool. If not, you may need tot add a generic parameter to this class to specifies which range of enum values are valid (so for the Chat reducer this might be `BaseMessage<CommonMessageType | ChatMessageType>` 

Aside from that I'll also note that the cast here made it so the type was explicitly equal to the enum value specified, so you couldn't e.g. initialize a TextMessageRecord with `type: 'whateverFakeType'`, it *had* to have the right type if you were using that Record. Is there some compilation problem you're trying to avoid by removing the type, or did you just think it was unnecessary?

Reply:
I had no idea what I was doing so I removed them since TS compiler didn't complain :p